### PR TITLE
Plugin custom save

### PR DESCRIPTION
**Pull request by ronwnor**
lil plugin to save graphs with custom thumbnails
image

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,3 +2,4 @@ node_modules
 dist
 *.zip
 coverage
+calculator.js
```
Comment (FenyLabs) [inline .gitignore:5]: Might want to remove this, probably just for development

File: src/plugins/custom-save/Controller.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+function updateView() {
+  desModderController.updateMenuView();
+}
+
+export default class Controller {
+  checked: Boolean;
+
+  constructor() {
+    this.checked = false;
```
Comment (FenyLabs) [inline src/plugins/custom-save/Controller.ts:11]: Tiny nitpick but you could probably use a default value here

File: src/plugins/custom-save/index.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  controller = new Controller();
+  return controller;
+}
+
+const CustomSave: Plugin = {
+  id: "custom-save",
+  onEnable,
+  onDisable: () => {},
+  onConfigChange: () => {
+    console.log("AAAAAAA");
```
Comment (FenyLabs) [inline src/plugins/custom-save/index.ts:16]: Again might want to remove before merge

File: src/preload/moduleOverrides/custom-save.replacements [hunk 4]
```diff
@@ -0,0 +1,77 @@
+# Custom Save Replacements
+
+*plugin* `custom-save`
+
+## Add a checkbox to the save dialog
+
+*module* `custom-save`
```
Comment (FenyLabs) [inline src/preload/moduleOverrides/custom-save.replacements:7]: Not sure if fireflame cares about this anymore, but I think you're overriding `calculator-shell/save-dialog`

File: src/preload/moduleReplacements.ts [hunk 5]
```diff
@@ -23,6 +24,7 @@ const replacementStrings = [
   debugMode,
   extraExpressionButtons,
   showTips,
+  CustomSave,
```
Comment (FenyLabs) [inline src/preload/moduleReplacements.ts:27]: camelCase instead of PascalCase for consistency?

File: src/plugins/custom-save/Controller.ts [hunk 6]
```diff
@@ -0,0 +1,22 @@
+import { desModderController } from "script";
+
+function updateView() {
+  desModderController.updateMenuView();
```
Comment (FenyLabs) [inline src/plugins/custom-save/Controller.ts:4]: Is this doing anything? Was under the impression this was to update DesModder menus/pillboxes, not Desmos's. Perhaps Calc.controller.updateViews()?

**Review by FenyLabs — CHANGES_REQUESTED** [review]
Just a couple tiny things to change to clean up. However, the plugin feels a bit awkward to use and depends on video creator output being a square. Not sure how to solve this, but perhaps the checkbox could be moved to be a part of video creator ui? or it could somehow be made more clear how to use the plugin. Currently it's not very clear that the rename modal also contains the checkbox, so perhaps that dropdown element could be renamed or something.

File: src/plugins/custom-save/index.ts (hunk 3 above)
Comment (ronwnor) [reply src/plugins/custom-save/index.ts:16]: ok this is embarrassing ///

File: .gitignore (hunk 1 above)
Comment (ronwnor) [reply .gitignore:5]: oh whoops

File: src/preload/moduleOverrides/custom-save.replacements (hunk 4 above)
Comment (jared-hughes) [reply src/preload/moduleOverrides/custom-save.replacements:7]: It's ok. The specific module names are pretty irrelevant now. They can probably be deleted throughout, but they're not doing much harm.

**Comment by jared-hughes** [comment]
It's OK if the video creator output is not a square, as long as the thumbnail gets cropped to be square.

Another consideration to the screenshot dimensions is if they're too big. Maybe downscale the image by default to a 400x400 max? Or maybe the best approach could be to warn the user if they're trying to upload a huge image, and give them an option to cancel.

For dynamic resize in JS (tweak the drawImage call).

```js
const canvas = document.createElement("canvas");
const ctx = canvas.getContext("2d");
ctx.drawImage(img, 0, 0, 300, 300);
const dataurl = canvas.toDataURL("image/png");
```

I almost never go into the "rename" menu, so I resonate with Feny's idea of moving the button to the video creator menu. It's easier to add a confirmation button before activating the native save-graph menu than to create a bunch of replacements.

**Comment by jared-hughes** [comment]
This would fix #68.

**Comment by jared-hughes** [comment]
Closing this because stale. If you want to restart it, I suggest going from main and re-applying changes.

**Pull request closed without merging**